File: Mobiles/ios/QuickPizzaIos/Features/Profile/Presentation/ProfileViewModel.swift
import SwiftUI
import SwiftiePod

let profileViewModelProvider = Provider(scope: AlwaysCreateNewScope()) { pod in
    ProfileViewModel(
        authService: pod.resolve(authRepositoryProvider),
        ratingsRepository: pod.resolve(ratingsRepositoryProvider),
        logger: pod.resolve(loggerProvider)
    )
}

@Observable
class ProfileViewModel {
    private let authService: AuthServiceProtocol
    private let ratingsRepository: RatingsRepositoryProtocol
    private let logger: Logging

    // View state
    var ratings: [Rating] = []
    var isLoading = false
    var errorMessage: String?
    var didLogout = false

    init(
        authService: AuthServiceProtocol,
        ratingsRepository: RatingsRepositoryProtocol,
        logger: Logging
    ) {
        self.authService = authService
        self.ratingsRepository = ratingsRepository
        self.logger = logger
    }

    var username: String {
        authService.currentUsername ?? "Unknown"
    }

    var pizzaCount: Int {
        ratings.count
    }

    /// Called from .task { } when the view appears.
    func start() async {
        await loadRatings()
    }

    func loadRatings() async {
        isLoading = true
        do {
            ratings = try await ratingsRepository.getRatings()
        } catch {
            logger.error("Failed to load ratings", error: error)
            errorMessage = "Failed to load ratings"
        }
        isLoading = false
    }

    func clearRatings() async {
        do {
            try await ratingsRepository.clearRatings()
            ratings = []
            logger.info("Ratings cleared by user")
        } catch {
            errorMessage = error.localizedDescription
            logger.error("Failed to clear ratings", error: error)
        }
    }

    func signOut() {
        authService.logout()
        didLogout = true
        logger.info("User signed out")
    }
}

File: Mobiles/android/app/src/main/java/com/grafana/quickpizza/features/profile/presentation/ProfileViewModel.kt
package com.grafana.quickpizza.features.profile.presentation

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.grafana.quickpizza.core.o11y.AppLogger
import com.grafana.quickpizza.features.auth.domain.AuthRepository
import com.grafana.quickpizza.features.profile.domain.RatingsRepository
import com.grafana.quickpizza.features.profile.models.Rating
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import javax.inject.Inject

data class ProfileUiState(
    val username: String? = null,
    val ratings: List<Rating> = emptyList(),
    val isLoading: Boolean = false,
    val errorMessage: String? = null,
    val showRatingsClearedMessage: Boolean = false,
)

@HiltViewModel
class ProfileViewModel @Inject constructor(
    private val ratingsRepository: RatingsRepository,
    private val authRepository: AuthRepository,
    private val logger: AppLogger,
) : ViewModel() {

    private val _state = MutableStateFlow(ProfileUiState())
    val state: StateFlow<ProfileUiState> = _state.asStateFlow()

    init {
        viewModelScope.launch {
            val name = withContext(Dispatchers.IO) { authRepository.username }
            _state.update { it.copy(username = name) }
        }
        loadRatings()
    }

    private fun loadRatings() {
        viewModelScope.launch {
            _state.update { it.copy(isLoading = true) }
            try {
                val ratings = ratingsRepository.getRatings()
                _state.update { it.copy(ratings = ratings) }
            } catch (e: Exception) {
                logger.error("Failed to load ratings", e)
                _state.update { it.copy(errorMessage = e.message) }
            } finally {
                _state.update { it.copy(isLoading = false) }
            }
        }
    }

    fun clearRatings() {
        viewModelScope.launch {
            try {
                ratingsRepository.clearRatings()
                _state.update { it.copy(ratings = emptyList(), showRatingsClearedMessage = true) }
                logger.info("Ratings cleared")
            } catch (e: Exception) {
                logger.error("Failed to clear ratings", e)
            }
        }
    }

    fun onRatingsClearedMessageShown() {
        _state.update { it.copy(showRatingsClearedMessage = false) }
    }

    fun signOut(onSignedOut: () -> Unit) {
        authRepository.logout()
        logger.info("User signed out")
        onSignedOut()
    }
}
